### ISSUE TITLE
aarch64 compile fails

### ISSUE THREAD

**Issue by gandalth** [yuzu-emu/yuzu#5290]
I am compiling on Ubuntu 20.10 (aarch64, RPi4) using command cmake .. -GNinja -DCMAKE_C_COMPILER=gcc-10 -DCMAKE_CXX_COMPILER=g++-10 and then invoking ninja. Compile fails with:

../src/core/arm/dynarmic/arm_dynarmic_64.cpp:7:10: fatal error: dynarmic/A64/a64.h: No such file or directory

The compilation continues when I manually add the corresponding directory to the includes list:
-I../externals/dynarmic/include/

I found this include directive to be missing for the following targets:

arm_dynarmic_32.cpp.o
arm_dynarmic_64.cpp.o
arm_dynarmic_cp15.cpp.o
arm_exclusive_monitor.cpp.o
physical_core.cpp.o

Finally, creation of physical_core.cpp.o fails with the following message:
../src/core/hle/kernel/physical_core.cpp:34:2: error: #error Platform not supported yet.

**Comment by NielsGx** [comment]
Platform not supported yet.

**Comment by gandalth** [comment]
Thank you for sharing!
Does this apply to ARM in general or just to ARM64?

**Comment by german77** [comment]
ARM is not supported

**Comment by Wooty-B** [comment]
Should have checked this out sooner... I'm compiling on a 16core ARM64 workstation with AMD WX 4100 GPU and want to confirm if you mean ARM isn't supported in general or if aarch64 could work? Here is where Ninja fails for me:

```
[395/1124] Building CXX object src/common/CMakeFiles/common.dir/wall_clock.cpp.o
FAILED: src/common/CMakeFiles/common.dir/wall_clock.cpp.o
...
In file included from ../src/common/wall_clock.cpp:7:
../src/./common/uint128.h:16:10: fatal error: x86intrin.h: No such file or directory
   16 | #include <x86intrin.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
```

More than happy to provide more info or test. Thanks!!!

**Comment by v1993** [comment]
ARM is not currently supported, but can be supported (it's not high priority for now since quite some important stuff is left to do). The biggest part of work would go into porting dynarmic (with potentially large performance increase due to less work on ARM-ARM emulation), but minor adjustments to emulator code also will be needed. It doesn't really matter what GPU you have if problem is during compilation btw.

**Comment by Wooty-B** [comment]
I see, I just see so many threads that ARM isn't powerful enough, however the ARM-ARM dynarmic impact is something I didn't think about. I appreciate the the time and will keep a close eye on this project!

**Comment by YLinXin** [comment]
> Should have checked this out sooner... I'm compiling on a 16core ARM64 workstation with AMD WX 4100 GPU and want to confirm if you mean ARM isn't supported in general or if aarch64 could work? Here is where Ninja fails for me:
> 
> ```
> [395/1124] Building CXX object src/common/CMakeFiles/common.dir/wall_clock.cpp.o
> FAILED: src/common/CMakeFiles/common.dir/wall_clock.cpp.o
> ...
> In file included from ../src/common/wall_clock.cpp:7:
> ../src/./common/uint128.h:16:10: fatal error: x86intrin.h: No such file or directory
>    16 | #include <x86intrin.h>
>       |          ^~~~~~~~~~~~~
> compilation terminated.
> ```
> 
> More than happy to provide more info or test. Thanks!!!

It should change x86intrin.h to arm_neon.h(even [arm_sve.h](https://developer.arm.com/documentation/102131/latest/) ) from [here](https://stackoverflow.com/questions/11228855/header-files-for-x86-simd-intrinsics#answer-22291538).
I can pass the include but more error shows:
```[632/1245] Building CXX object src/core/CMakeFiles/core.dir/arm/dynarmic/arm_dynarmic_32.cpp.o
FAILED: src/core/CMakeFiles/core.dir/arm/dynarmic/arm_dynarmic_32.cpp.o 
/usr/bin/g++-10 -DARCHITECTURE_ARM64=1 -DBOOST_ALL_NO_LIB -DBOOST_ASIO_DISABLE_CONCEPTS -DBOOST_CONTEXT_DYN_LINK -DBOOST_DATE_TIME_NO_LIB -DBOOST_ERROR_CODE_HEADER_ONLY -DBOOST_REGEX_NO_LIB -DBOOST_SYSTEM_NO_LIB -DENABLE_WEB_SERVICE -DFMT_LOCALE -DFMT_SHARED -DJSON_DIAGNOSTICS=0 -DJSON_USE_IMPLICIT_CONVERSIONS=1 -DNDEBUG -DOPUS_VERSION="\"\\\"1.3.1\\\"\"" -DYUZU_UNIX=1 -D_FILE_OFFSET_BITS=64 -I/home/ubuntu/yuzu/src/. -I/home/ubuntu/yuzu/externals/./microprofile -I/home/ubuntu/yuzu/externals/glad/include -I/home/ubuntu/yuzu/externals/sirit/src/../include -I/home/ubuntu/yuzu/externals/sirit/externals/SPIRV-Headers/include -I/home/ubuntu/yuzu/externals/./mbedtls/include -I/home/ubuntu/yuzu/externals/opus/opus/include -O3 -DNDEBUG -Wall -Werror=array-bounds -Werror=implicit-fallthrough -Werror=missing-declarations -Werror=missing-field-initializers -Werror=reorder -Werror=sign-compare -Werror=switch -Werror=uninitialized -Werror=unused-function -Werror=unused-result -Werror=unused-variable -Wextra -Wmissing-declarations -Wno-attributes -Wno-invalid-offsetof -Wno-unused-parameter -fconcepts -Werror=conversion -Werror=ignored-qualifiers -Werror=shadow -Werror=class-memaccess -Werror=unused-but-set-parameter -Werror=unused-but-set-variable -Wno-sign-conversion -std=gnu++2a -MD -MT src/core/CMakeFiles/core.dir/arm/dynarmic/arm_dynarmic_32.cpp.o -MF src/core/CMakeFiles/core.dir/arm/dynarmic/arm_dynarmic_32.cpp.o.d -o src/core/CMakeFiles/core.dir/arm/dynarmic/arm_dynarmic_32.cpp.o -c /home/ubuntu/yuzu/src/core/arm/dynarmic/arm_dynarmic_32.cpp
/home/ubuntu/yuzu/src/core/arm/dynarmic/arm_dynarmic_32.cpp:7:10: fatal error: dynarmic/interface/A32/a32.h: No such file or directory
    7 | #include <dynarmic/interface/A32/a32.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[641/1245] Building CXX object src/shader_recompiler/CMakeFiles/shader_recompiler.dir/frontend/maxwell/translate/impl/integer_add_three_input.cpp.o
ninja: build stopped: subcommand failed.
```

**Comment by YLinXin** [comment]
> I see, I just see so many threads that ARM isn't powerful enough, however, the ARM-ARM dynamic impact is something I didn't think about. I appreciate the time and will keep a close eye on this project!

ARM is powerful enough to do this in servers and supercomputers and apple supplies the most powerful M1 family powered by ARM to the custom markets. There is no need for instruction translate and something is [more power on ARM from dolphin](https://dolphin-emu.org/blog/2021/05/24/temptation-of-the-apple-dolphin-on-macos-m1/?nocr=true). So I think we should put ARM to Tier 1.

**Comment by v1993** [comment]
Despite some servers and supercomputers using ARM, single-thread performance of it is still far from perfect. And when it comes to consumer-available chips you end up with just M1, which in turn has pretty bad situation with GPU drivers and limited userbase. So while ARM support in long term would be great, there are still far more important tasks for now. In addition, appearance of powerful ARM processors for other platforms will ease the development considerably, since you won't have to figure out various quirks of OS X and moltenvk while working on CPU emulation code.

TL;DR: ARM is looking promising, but it's just not quite there yet.

**Comment by YLinXin** [comment]
> Despite some servers and supercomputers using ARM, single-thread performance of it is still far from perfect. And when it comes to consumer-available chips you end up with just M1, which in turn has pretty bad situation with GPU drivers and limited userbase. So while ARM support in long term would be great, there are still far more important tasks for now. In addition, appearance of powerful ARM processors for other platforms will ease the development considerably, since you won't have to figure out various quirks of OS X and moltenvk while working on CPU emulation code.
> 
> TL;DR: ARM is looking promising, but it's just not quite there yet.

I agree, the best consumer ARM chip, the worst dev-friendly OS.
apple fucker.

**Comment by v1993** [comment]
As far as M1 linux goes - situation is even worse on it with GPU drivers than in OS X for now. A lot more progress in that regard is required before it will have features needed for yuzu.

**Comment by YLinXin** [comment]
> As far as M1 linux goes - situation is even worse on it with GPU drivers than in OS X for now. A lot more progress in that regard is required before it will have features needed for yuzu.

eieiei, I build yuzu in the virtual machine Linux, not in physical Linux installed on M1. I just want to test compile status on arm.
there will be more and more arm Linux devices in the market, especially in China market.

**Comment by Wooty-B** [comment]
> > As far as M1 linux goes - situation is even worse on it with GPU drivers than in OS X for now. A lot more progress in that regard is required before it will have features needed for yuzu.
> 
> eieiei, I build yuzu in the virtual machine Linux, not in physical Linux installed on M1. I just want to test compile status on arm. there will be more and more arm Linux devices in the market, especially in China market.

I agree on this front as I have a 16core w/ up to 2.5Ghz cores and a 4GB AMD Radeon attached with working GPU drivers. Even if the performance was slow, I would be interested in just tinkering with it in the meantime. Emulators like Dolphin and Citra run at native speeds with most games at 2-3x upscale on my system, as well as games like Doom 3, and Minecraft @ 16 chunks.

Again, I understand this is focused more towards x86 at the moment, however ARM is a growing platform for many open systems as well as consumer devices. If theres anything I can do to help test on ARM I am willing. Regardless, thanks for your time and work on this project!!

**Comment by neofeo** [comment]
The swtich has very poor single core performance, even a cortex a72 at 2.0/2.2 ghz should  be able to run an eventual yuzu for arm64, while the main constraints remains being the gpu. Even latest mali GPUs will be very constrained, taking aside that the mali blobs tend to not perform great on linux, and that panfrost has still a long way to reach latest mali gpus (an year, to say at least). Panfrost landed on g57 tho. No idea about the performance. We can expect decent gl es performance on blobs anyway, but I am mostly sure that the switch games use pure opengl.. if they use GLES, then we should  be fine with the blobs across all platforms.

Adreno gpus have great mesa drivers, vulkan include. Imagination gpus are also being supported at mesa, vulkan include. 

Don't know thst much about the emulator, but if it use vulkan or gl es, specially the second one, then yuzu should be a great deal for handhelds on arm linux, and of course, on android

**Comment by jbeich** [comment]
Can you re-try after https://github.com/yuzu-emu/yuzu/pull/9198 ?
